### PR TITLE
Fix issue with cache

### PR DESCRIPTION
**Pull request by Helium314**
see https://github.com/streetcomplete/StreetComplete/issues/4980#issuecomment-1531960544

No tests yet, so far I failed to reproduce the bug in a unit test.

### CODE REVIEW

**Comment by westnordost** [comment]
TBH I understand neither the problem nor the solution. But this is not a requirement, unless you need feedback from me.

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt [hunk 1]
```diff
@@ -383,8 +386,13 @@ class MapDataCache(
                 return result
             }
 
-            // get nodes again, this contains the newly added nodes, but maybe not the old ones if cache was trimmed
-            nodes.addAll(spatialCache.get(bbox))
+            // add newly fetched nodes from elements
+            // getting nodes from spatialCache can cause issues, as tiles in the bbox may now be removed unexpectedly
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt:390]: how/when though? This is all synchronized, isn't it?

**Comment by Helium314** [comment]
What happens:
Tile 1 is cached from a previous fetch, then `getMapDataWithGeometry` is called for a bbox (exactly) containing tile 1 and _adjacent_ tile 2.
[`tilesToFetch`](https://github.com/streetcomplete/StreetComplete/blob/59e44eaae360181eb649df19b63ae775d4d8476d/app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt#L360) will contain tile 2 only, because tile 1 is already cached.
In `nodes = HashSet<Node>(spatialCache.get(bbox))`, the `SpatialCache` will use [its own `fetch`](https://github.com/streetcomplete/StreetComplete/blob/59e44eaae360181eb649df19b63ae775d4d8476d/app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt#L117-L120), but `SpatialCache.fetch` [always returns an empty list](https://github.com/streetcomplete/StreetComplete/blob/59e44eaae360181eb649df19b63ae775d4d8476d/app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt#L29) for tile 2. This adds tile 2 to the `SpatialCache`, containing the empty list returned by `SpatialCache.fetch`.
At this point there is no issue, because the tile being empty is expected, and we replace the fetched empty tile 2 with the actual data as part of the [immediately following `update`](https://github.com/streetcomplete/StreetComplete/blob/59e44eaae360181eb649df19b63ae775d4d8476d/app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt#L64).
But during the `update`, the `tilesRect.asBoundingBox(zoom).enclosingTilesRect(zoom)` bug(?) sometimes kicks in, and the unexpectedly [extended larger tiles rect](https://github.com/streetcomplete/StreetComplete/blob/59e44eaae360181eb649df19b63ae775d4d8476d/app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt#L86) suddenly contains tile 1 and tile 2, instead of tile 2 only. Tile 1 is then seen as [incomplete](https://github.com/streetcomplete/StreetComplete/blob/59e44eaae360181eb649df19b63ae775d4d8476d/app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt#L87), because it's not fully in the bbox that was supposed to contain tile 2 only.
Since tile 1 is supposed to be replaced with incomplete data, tile 1 is removed from `SpatialCache`.

Next, `MapDataCache` fetches the nodes added to `SpatialCache` in `update`, and for shorter code it was done using [`SpatialCache.get`](https://github.com/streetcomplete/StreetComplete/blob/59e44eaae360181eb649df19b63ae775d4d8476d/app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt#L387). I had seen this as safe, but now we know it's not.
As described above, when now calling `get` for the full bbox of tile 1 and tile 2, the previously removed tile 1 is fetched using `SpatialCache.fetch`, so it contains no nodes. But this time `MapDataCache` isn't aware of this, and the empty tile is seen as properly cached. On this fetch, the returned data is still correct. On further fetches in tile 1, nothing will be returned.

With this PR, the second `SpatialCache.fetch` is not done, and thus if the `tilesRect.asBoundingBox(zoom).enclosingTilesRect(zoom)` bug occurs, the removed tile 1 is simply fetched again from DB.

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt (hunk 1 above)
Comment (Helium314) [reply app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt:390]: Updating the cache may remove tiles, see comment below for an explanation.

**Comment by Helium314** [comment]
added a test that fails without the change to `MapDataCache`

**Comment by Helium314** [comment]
I'll merge this, as it works, and with the PR merged I'll continue #4980 (might actually be finished already, assuming the issues I observed were really caused by this bug only).

**Comment by westnordost** [comment]
Thank you for the explanation. 

It makes sense to not rely on `tilesRect.asBoundingBox(zoom).enclosingTilesRect(zoom) == tilesRect` because when one data structure is inclusive and the other exclusive to its bound, it is somewhat difficult or impossible to have a lossless conversion between the two.

**Pull request merged**